src/main/java/io/supertokens/jwt/JWTSigningKey.java

```diff
@@ -0,0 +1,182 @@
+
+            return sqlStorage.startTransaction(con -> {
+               List<JWTSigningKeyInfo> keys;
+               List<JWTSigningKeyInfo> keysFromStorage = sqlStorage.getJWTSigningKeys_Transaction(con);
+
+               if (keysFromStorage == null) {
+                   keys = new ArrayList<>();
+               } else {
+                   keys = keysFromStorage;
+               }
+
+               sqlStorage.commitTransaction(con);
+               return keys;
+            });
+        } else if (storage.getType() == STORAGE_TYPE.NOSQL_1) {
+            SessionNoSQLStorage_1 noSQLStorage = (SessionNoSQLStorage_1) storage;
+
+            List<JWTSigningKeyInfo> keys;
+            List<JWTSigningKeyInfo> keysFromStorage = noSQLStorage.getJWTSigningKeys_Transaction();
+
+            if (keysFromStorage == null) {
+                keys = new ArrayList<>();
+            } else {
+                keys = keysFromStorage;
+            }
+
+            return keys;
+        }
+
+        throw new QuitProgramException("Unsupported storage type detected");
+    }
+
+    public synchronized JWTSigningKeyInfo getKeyForAlgorithm(String algorithm)
+            throws NoSuchAlgorithmException, StorageQueryException, StorageTransactionLogicException {
+        if (!JWTSigningKey.isJWTAlgorithmSupported(algorithm)) {
+            throw new NoSuchAlgorithmException(algorithm + "is not a supported signing algorithm");
+        }
+
+        SessionStorage storage = StorageLayer.getSessionStorage(main);
+
+        if (storage.getType() == STORAGE_TYPE.SQL) {
+            SessionSQLStorage sqlStorage = (SessionSQLStorage) storage;
+
+            return sqlStorage.startTransaction(con -> {
+               JWTSigningKeyInfo keyInfo;
+               String algorithmType;
+
+               try {
+                   algorithmType = getAlgorithmType(algorithm);
+               } catch (NoSuchAlgorithmException e) {
+                   throw new StorageTransactionLogicException(e);
+               }
+
+               JWTSigningKeyInfo keyFromStorage = sqlStorage.getLatestJWTSigningKeyForAlgorithm_Transaction(con, algorithmType);
+
+               if (keyFromStorage == null) {
+                   try {
+                       long currentTimeInMillis = System.currentTimeMillis();
+                       Utils.PubPriKey newKey = Utils.generateNewPubPriKey();
+                       keyInfo = new JWTSigningKeyInfo(Utils.getUUID(), newKey.publicKey, newKey.privateKey, currentTimeInMillis, algorithm, algorithmType);
+                       sqlStorage.setJWTSigningKey_Transaction(con, keyInfo);
+                   } catch (NoSuchAlgorithmException e) {
+                       throw new StorageTransactionLogicException(e);
+                   }
+               } else {
+                   keyInfo = keyFromStorage;
+               }
+
+               sqlStorage.commitTransaction(con);
+               return keyInfo;
+            });
```

Comment:
Done, since NoSuchAlgorithm is an exception that Java's cryptography throws I am now using a custom UnsupportedAlgorithmException for our use case (the names do seem confusing, maybe I should call it something else?)

Reply:
The exception is now called `UnsupportedJWTSigningAlgorithmException`